openvinotoolkit/nncf
# ONNX Post-Training MinMax Quantization

### Changes

Introduce MinMax Post-Training Quantization algorithm for ONNX:

1. Add **post-training** section, which consists of the framework-agnostic classes and methods that are essential for enabling post-training algorithms.

The main classes are:
```
CompressionBuilder
StatisticsAggegator
DataLoader
Engine
Sampler
Algorithm
```


2. Add implementation of post-training section for ONNX backend:
    1. Metatypes: Add the mapping between ONNX ops and hardware ops.
    2. Add two algorithms: ONNXPostTrainingQuantization, ONNXMinMaxQuantization.
    4. Add ONNXModelTransformer.
    5. Add ONNXInsertionCommands: ONNXQuantizerInsertionCommand
    6. Update NNCFGraphBuilder for ONNX, as well as ONNXGraph.
    7. Add ONNXStatisticsAggegator.
    8. Add ONNXNNCFTensor as well as ONNXNNCFCollectorTensorProcessor.

TODO:

- [x] Add test on StatisticsAggegator
- [x] Add test on Sampler


### Reason for changes

65600

### Related tickets

76191

### Tests

1. Add test on models graph after ONNXMinMaxQuantization and PostTrainingQuantization were added for 
- resnet18
- mobilenet_v2
- inception_v3
- googlenet
- vgg16
- shufflenet_v2_x1_0
2. Add test on ONNXModelTransformer -  tries to insert quantizers to the synthetic model and then check their numbers in the obtained model. 
3. Add test on ONNXModelTransformer -  tries to insert quantizers to the synthetic model and then check their parameters in the obtained model. 
4. Add test on ONNXBatchSampler and ONNXRandomBatchSampler - check the samples with the targets with the reference for different batch_size

**What should be done in the next PRs:**

1. Add saturation fix support
2. Add a model inference step to get the edges shape of the model for the models that don't have the shapes.
3. Add PTQ sample for ONNX classification with the sanity test on the sample.
